[14.0][MIG] l10n_br_account_nfe

antoniospneto opened the pull request:
Migração do módulo **l10n_br_account_nfe**

O maior problema que eu enfrentei foi Inconsistências relacionado a company: 

Movi os dados de demostração para que fosse carregado pelo hook, devido a varias inconsistências relacionado ao company_id. Nos dados de demostração temos uma fatura da empresa **Simples Nacional**, porém quando é carregado da forma tradicional, informando o arquivo demo no manifest, o arquivo é carregado com a empresa padrão do usuário (OdooBot) setada na empresa **"Sua Empresa".**  Isso era um problema, pois apesar da invoice ter o campo company_id e o mesmo estar explicitamente informado nos dados demo, essa informação não era herdada para o inherits do documento fiscal, devido a limitação da ORM quando existe os campos com o mesmo nome, apesar de existir a logica dos  Shadow fields da nossa localização para fazer a correção destes casos, essa correção é feita apenas depois do método create(), porém agora no Odoo 14.0 muita logica acontece dentro do create, muita mesmo rss, nesse tempo a company da invoice ficava diferente do inherits do documento fiscal e essas informações da company era replicada para as linhas.. Além disso outro grande problema era na logica do fiscal dummy, pois ele pega como referencia a company padrão do usuário no momento (self.env.company) que ficava diferente da company da invoice.. isso fazia carregar linhas ficais dummy onde não deveria..

Ao mover o carregamento do demo pelo hooks todos esses problemas são resolvidos pois lá eu altero a empresa padrão do usuário para a empresa "Simples Nacional" antes do carregamento. dessa forma não há chance de algo ser carregado como sendo da empresa "Sua Empresa".

Outra coisa que eu fiz foi adicionar a verificação de inconsistência da company no documento fiscal, ver commit: ea888e89db70e9681f03cfb2cd186c365a616b56 isso vai economizar bastante tempo de debug se alguém cair no mesmo problema hehe.

felipemotter committed:
[ADD] new module l10n_br_account_nfe

antoniospneto committed:
add cobrança info nfe

antoniospneto committed:
add cobrança test

antoniospneto committed:
[FIX] generate financial information only when the issuer of nfe is a company

mbcosta committed:
[IMP] Demo Data.

mbcosta committed:
[REF] Changed the object used to create NFe Tags pag and dup because only when the creation are done in l10n_br_fiscal.document the Tags appear in XML.

mbcosta committed:
[REF] Rename file after changed the object used to create NFe Tags PAG e DUP.

mbcosta committed:
[REF+IMP] Changed Tests to use SavepointCase and included test with demo data.

mbcosta committed:
[REF] Name of test file.

mbcosta committed:
[UPD] Readme files.

mbcosta committed:
[REF+FIX] Included License Header, made module auto install and fix module version.

oca-travis committed:
[UPD] Update l10n_br_account_nfe.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

renatonlima committed:
Update financial move in fiscal document NFE XML

renatonlima committed:
[FIX] clean fields nfe40_dup and nfe40_detPag before create records

oca-travis committed:
[UPD] Update l10n_br_account_nfe.pot

renatonlima committed:
Não deve criar nfe40_dup e record.nfe40_detPag quando a finalidade for devolução

OCA-git-bot committed:
l10n_br_account_nfe 12.0.1.1.0

oca-travis committed:
[UPD] Update l10n_br_account_nfe.pot

OCA-git-bot committed:
l10n_br_account_nfe 12.0.1.1.1

renatonlima committed:
l10n_br_account_nfe: pre-commit after pre-commit version upgrade

antoniospneto committed:
[MIG] l10n_br_account_nfe: Migration to 14.0

antoniospneto committed:
[IMP] l10n_br_account_nfe: pre-commit execution

antoniospneto committed:
[MIG] rename fields

antoniospneto committed:
[MIG] load the demo in the hook

antoniospneto committed:
[MIG] fix tests

antoniospneto committed:
[IMP] add check company for document fiscal

antoniospneto committed:
[FIX] dont set None in float

antoniospneto committed:
[FIX] company consistency payment order demo

antoniospneto committed:
[FIX] back default user company to main_company

antoniospneto commented:
Alterei o hook para que no fim da execução a empresa padrão do usuário voltasse para empresa "Sua Empresa" (main_company) pois estava bagunçando a instalação dos outros módulos, eles esperam que essa seja a empresa padrão.

@rvalyi o erro que ainda está acontecendo no Travis é aquela questão relacionado ao spec_driven_model que nesse momento a injeção dos mixing da nfe ainda não foram totalmente processados. localmente eu consigo contornar isso fazendo primeiro a instalação do módulo l10n_br_nfe, reinicio o Odoo e então faço a instalação do l10n_br_account_nfe com os testes ativos, se fizer nessa ordem não acontece o erro.
```
File "/home/travis/odoo-14.0/odoo/models.py", line 3011, in read
raise ValueError("Invalid field %r on model %r" % (name, self._name))
ValueError: Invalid field 'nfe40_fat' on model 'l10n_br_fiscal.document'
```
Tem alguma sugestão de como podemos resolver ?

rvalyi commented:
ola @netosjb eu nao tenho como mergulhar nisso 100% esses dias mas vou fazer o possivel para ajudar. Entao, na 12.0, se vc fizer `odoo -d db -i l10n_br_account_nfe --stop-after-init --test-enable`
vc tambem pega um erro https://gist.github.com/rvalyi/7c3800c3bff88d44a93351bd5d754775
(Se alguem quiser tentar, aconselho instalar pelo menos ate o account sem os tests antes senão leva a vida.)

O Travis e Github Actions realmente instalam os modulos ANTES de rodar os testes. Isso é para facilitar a "idempotence" dos testes. Assim localmente se algo falhar vc pode tentar resolver e rodar os testes de novo, nao precisa re-instalar tudo do zero...

Outra informação: na 12.0 também o comportamento é de carregar os dados de demo ANTES de rodar o post_init_hook, não houve mudança na v14 nisso.

Ai realmente com o framework spec_model_driven depois de instalar todos os modulos é chamado o `register_hook` que frisa o mapping [https://github.com/OCA/l10n-brazil/blob/14.0/spec_driven_model/hooks.py#L15](https://github.com/OCA/l10n-brazil/blob/14.0/spec_driven_model/hooks.py#L126). Eu fiz esse design para nao perder a modularidade do Odoo: assim vc poderia ter outros modulos de NFe que podem mapear alguns campos em outros objetos comparando com uma instalação basica. Mas realmente poderia levar a surpresas dependente do tipo de dado de demo que vc ta tentando carregar ante desse hook ser chamado. Por isso no l10n_br_nfe eu tento importar NFe's de demo apenas DEPOIS de chamar o `register_hook` https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_nfe/hooks.py#L19 pois se espera que o mapeando dos objetos de NFe estejam totalmente mapeados.

Espero que isso clarifica algumas coisas. Dito isso vou analisar um pouco melhor o que vc fez e tentar ajudar mais...

rvalyi commented:
@netosjb segura ai porque eu tenho coisas para subir, ate na 12.0 que deve ajudar. Vamos ver se eu resolvo hoje ainda...

rvalyi committed:
[IMP] load COAs if demo/test context

rvalyi committed:
[FIX] register_hook guard fix + cleanup

rvalyi committed:
[FIX] more robust spec ancestor injection

Indeed if register_hook is forced in another module were
some object is inherited again (example in l10n_br_account_nfe),
you won't find the spec mixins in _inherit and you need to look at mro()

rvalyi commented:
> Alterei o hook para que no fim da execução a empresa padrão do usuário voltasse para empresa "Sua Empresa" (main_company) pois estava bagunçando a instalação dos outros módulos, eles esperam que essa seja a empresa padrão.
> 
> @rvalyi o erro que ainda está acontecendo no Travis é aquela questão relacionado ao spec_driven_model que nesse momento a injeção dos mixing da nfe ainda não foram totalmente processados. localmente eu consigo contornar isso fazendo primeiro a instalação do módulo l10n_br_nfe, reinicio o Odoo e então faço a instalação do l10n_br_account_nfe com os testes ativos, se fizer nessa ordem não acontece o erro.
> 
> ```
> File "/home/travis/odoo-14.0/odoo/models.py", line 3011, in read
> raise ValueError("Invalid field %r on model %r" % (name, self._name))
> ValueError: Invalid field 'nfe40_fat' on model 'l10n_br_fiscal.document'
> ```
> 
> Tem alguma sugestão de como podemos resolver ?

na v12, mesmo depois de https://github.com/OCA/l10n-brazil/pull/2015 eu pego esse problema. Mas na 12 é apenas um WARNING. O Ponto é que nfe40_fat é um m2o que foi denormalizado na Nfe (stacked) e nisso o SpecModel tira o campo m2o na lista dos campos do objetos (e ele fica na lista dos stacking_points do objetos). Nisso ele nao deveria ser lido, tem que descobrir o que ta chamando o read desse campo...

rvalyi commented:
> @netosjb segura ai porque eu tenho coisas para subir, ate na 12.0 que deve ajudar. Vamos ver se eu resolvo hoje ainda...

segue o PR na 12.0 https://github.com/OCA/l10n-brazil/pull/2015

antoniospneto committed:
[FIX] hook for without demo

antoniospneto commented:
Update:

1. Adicionei os commit do @rvalyi  #2015 
2. Adicionei um fix que impedia a instalação quando fosse sem dados demo.

rvalyi commented:
@netosjb eu devo ter feito algo errado porque eu nao consegui reproduzir esse erro do nfe40_fat com sua branch na v14...
Mas testa esse fix:

```
diff --git a/l10n_br_nfe/models/document.py b/l10n_br_nfe/models/document.py
index b2a6b80745..46f76d9a2d 100644
--- a/l10n_br_nfe/models/document.py
+++ b/l10n_br_nfe/models/document.py
@@ -535,6 +535,7 @@ class NFe(spec_models.StackedModel):
                 comodel = self.env[self._stacking_points.get(field_name).comodel_name]
                 fields = [
                     f for f in comodel._fields if f.startswith(self._field_prefix)
+                    and f in self._fields.keys()
                 ]
                 sub_tag_read = self.read(fields)[0]
                 if not any(

```

me pareceria algo razoável.

rvalyi committed:
[FIX] export fields

antoniospneto commented:
aeeeee deu certo :D mandou bem @rvalyi , vou só arrumar o commit

rvalyi commented:
Basicamente no SpecModel eu removo os campos m2o quando os objetos ligados por esse campo são "stacked"/denormalizados para não ter eles poluindo. Nesse caso era só checar se o campo do comodel ainda tava no objeto ou não. Agora talvez eu me embananei nos testes aqui mas me parece que o pre_init_hook que tava puxando os módulos de COAs na 12 não tá puxando eles na 14. Mas testei rapidinho talvez eu fiz alguma besteira. Enfim só um toque mesmo para a gente verificar.

rvalyi commented:
Pessoal, me parece que o hook que eu fiz na 12 a semana passada não está puxando os COAs antes da instalação do l10n_br_account na 14. Alguém testou fazendo apenas -u l10n_account_nfe? Talvez eu que testei mal, mas eu acho que deu ruim de novo ontem no meu teste...

antoniospneto commented:
@rvalyi vou testar logo a noite 👍🏻

marcelsavegnago commented:
Em uma instancia na 14 estou com o mesmo problema. Vou tentar reproduzir em outra instancia. 
Obs. Até onde me lembro não havia instalado o l10n_br_account_nfe nesta instancia

Traceback (most recent call last):
  File "/home/odoo/prod/core/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/prod/core/odoo/http.py", line 316, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: Invalid field 'nfe40_fat' on model 'l10n_br_fiscal.document'

cc @augustodinizl

antoniospneto commented:
@marcelsavegnago, faça o teste com esse commit do @rvalyi , deve resolver: https://github.com/OCA/l10n-brazil/pull/2005/commits/adf0823c986aa2afd81cc19e7f7e28073dee3e86

marcelsavegnago commented:
> @marcelsavegnago, faça o teste com esse commit do @rvalyi , deve resolver: [adf0823](https://github.com/OCA/l10n-brazil/commit/adf0823c986aa2afd81cc19e7f7e28073dee3e86)

Funcionou... vlwwww.. vlwww @rvalyi

rvalyi commented:
@netosjb vc poderia fazer um rebase por favor? Sem os warning ficaria mais facil ver o que ta acontescendo com os hooks...

antoniospneto commented:
> Pessoal, me parece que o hook que eu fiz na 12 a semana passada não está puxando os COAs antes da instalação do l10n_br_account na 14. Alguém testou fazendo apenas -u l10n_account_nfe? Talvez eu que testei mal, mas eu acho que deu ruim de novo ontem no meu teste...

@rvalyi eu testei aqui, mas realmente o coa não está sendo instalado automaticamente :/

rvalyi reviewed (approved):
@netosjb @marcelsavegnago @renatonlima sobre o hook do l10n_br_account que nao instala os COAs eu quebrei a cabeça horas e horas mas eu ainda nao solucionei. Basicamente se vc rodar num banco virgem com -i l10n_br_account_nfe, os demos e testes quebrariam porque os modulos l10n_br_coa_simple e l10_br_coa_generic so seriam instalados depois. Mas se instalar em varias etapas ou sem demo ou sem test, nao tem esse problema (assim como na CI).

Eu entendi porque o hook não tentava instalar os COAs: é porque os modulos estão com o state "to install" e não "uninstalled" no caso que eu mencionei. Na v12 era "uninstalled". Se mudar a condição para pegar o state "to install" tb é forçar a instalação no pre_init_hook do l10n_br_account acontecem outros problemas com a intilializão dos campos dos modelos nesse pre_init_hook.

Com tudo eu acho que vamos ter que quebrar a cabeça varias horas ainda para arrumar esse hook do l10n_br_account para facilitar qdo tem demo ou testes.

Mas eu não acho que vale a pena segurar o merge do modulo l10n_br_account_nfe por conta disso, pois nas situações comuns (e ate para produção) ele ja funciona e vai facilitar as colaborações de fazer o merge.

rvalyi commented:
/ocabot merge nobump

rvalyi commented:
que louco, parece que o Pedro ficou trabalhando exatamente nesse problema de conseguir instalar imediatamente modulos com as dependencias. Ta no OpenUpgrade, mas talvez ajuda a solucionar esse hook do l10n_br_account la na frente...

https://github.com/OCA/OpenUpgrade/pull/3477
